runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/Work.java

```diff
@@ -171,6 +181,10 @@ public State getState() {
     return currentState.state();
   }
 
+  public @Nullable GetDataStream getDataStream() {
```

Comment:
as mentioned in the other change, could this be an interface restricted to heartbeats intead of full GetDataStream?
GetDataStream is confusing since it overlaps with getKeyedDataFn

Having it be some interface, also seems like we could possibly share more heartbeat logic between the paths as in general it could be
group work by HeartbeatRefreshBatcher (or better name)
call batcher.SendHeartbeats(work for that batcher)

and internally that could delegate to single GetDataStream for the direct case, or for SE could lookup a data stream from stub pool each time, or for SA could do nothing (or we set the batcher to null to do that).

Reply:
done

moved all functionality of heartbeats to the new HeartbeatSender interface
we can group by HeartbeatSender
dispatched SE and SA will have only 1 HeartbeatSender so these requests will all go to the same endpoint/receiver
direct path will fan out since there will be multiple HeartbeatSenders in the heartbeat request